### PR TITLE
[SPARK-36676][SQL][BUILD] Create shaded Hive module and upgrade Guava version to 30.1.1-jre

### PR DESCRIPTION
**Pull request by sunchao**
### What changes were proposed in this pull request?

This PR creates a new module `hive-shaded` which shades & relocates various dependencies from Hive, including Guava. By this means it also upgrades the Guava version from 14.0.1 to 30.1.1-jre.

### Why are the changes needed?

Spark currently use `hive-exec-core` which leaks a lot of dependencies to Spark, in particular Guava. As consequence, Spark is stuck with an ancient Guava version 14.0.1 which also carries CVE issues described in [SPARK-32502](https://issues.apache.org/jira/browse/SPARK-32502). 

By creating a shaded module, Spark is able to de-couple from those dependencies leaked by Hive and upgrade to newer versions of Guava. This also allows us to upgrade other dependencies whenever necessary, without having to wait for new Hive releases.

### Does this PR introduce _any_ user-facing change?

No. For Spark users who programmatically depending on modules such as `spark-hive`, the Hive dependencies will be replaced by the newly created `hive-shaded` module.

### How was this patch tested?

Existing tests.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Thank you, @sunchao .

cc @cloud-fan , @HyukjinKwon , @gengliangwang , @wangyum

File: dev/deps/spark-deps-hadoop-2.7-hive-2.3
Comment (dbtsai) [inline dev/deps/spark-deps-hadoop-2.7-hive-2.3:62]: Since we use shaded guava now in both hive, hadoop, spark, can we remove this dep entirely?

Comment (sunchao) [reply dev/deps/spark-deps-hadoop-2.7-hive-2.3:62]: It actually might be possible: seems Spark now shade Guava usage everywhere within itself now. I remember seeing a few modules which are not using shaded version but now can't find them anymore. Perhaps some mistake from my side.

It's a bit strange that Spark packs the shaded Guava classes in its `spark-network-common` jar. This is working since it is dependent by `spark-core` and thus transitively dependent by other modules.

I need to fix the PR since the "inherited" tag will prevent Spark from shading jetty & guava in other modules, which is bad. Thinking about how to do proper plugin inheritance in Maven...

**Review by dbtsai — APPROVED** [review]
LGTM. Thanks!

**Comment by pan3793** [comment]
If I understand correctly, the `hive-shaded` module creates a fat jar that contains all hive classes and the classes from transitive dependencies, but only a few classes are relocated. (Correct me if I'm wrong)

That means those classes which are not relocated will cause class conflict if the user also provides the original jars, e.g. HikariCP.

For this case, maybe we can learn something from https://github.com/trinodb/trino-hive-apache

**Comment by sunchao** [comment]
@pan3793 yes that's correct, I'm not relocating all the dependencies because the others haven't caused any issue so far AFAIK (after all, Spark today already leaks these dependencies). We can definitely include them if necessary, just need to check if they are used in APIs and whether that causes any breakage.

**Comment by pan3793** [comment]
> Spark today already leaks these dependencies

Yeah, but users can see what jars in `$SPARK_HOME/jars` and 1) align dependencies' version with Spark when build Spark application, 2) exclude jars already under `$SPARK_HOME/jars` when package or submit Spark job.

For `hive-shaded`, the user has no idea what jars should be aligned and exclude.

I think `hive-shaded` should relocate all classes except `org.apache.hive`, which makes things simple, the user just needs to exclude all hive vanilla jars.

**Comment by sunchao** [comment]
Makes sense. Let me try to relocate those too.

File: dev/deps/spark-deps-hadoop-2.7-hive-2.3
Comment (pan3793) [reply dev/deps/spark-deps-hadoop-2.7-hive-2.3]: Curator has a hard dependency on Guava, see detail at https://cwiki.apache.org/confluence/display/CURATOR/TN13

Comment (sunchao) [reply dev/deps/spark-deps-hadoop-2.7-hive-2.3]: Hmm you are right, I missed this. Somehow `mvn dependency:tree` doesn't show curator's guava dependency.

Comment (sunchao) [reply dev/deps/spark-deps-hadoop-2.7-hive-2.3]: So `build/sbt dependencyTree` provides a better result, and it looks like only `curator-client` and `hive-exec` has hard dependency on Guava at the moment. With the `hive-shaded` module, the latter is resolved so we are only left with the former, which makes me thinking whether we can just skip relocating the following 3 classes required by curator:

- com.google.common.base.Function
- com.google.common.base.Predicate
- com.google.common.reflect.TypeToken

It's not ideal but is better than exposing all the Guava classes.

Comment (pan3793) [reply dev/deps/spark-deps-hadoop-2.7-hive-2.3]: https://guava.dev/

> APIs without @Beta will remain binary-compatible for the indefinite future. (Previously, we sometimes removed such APIs after a deprecation period. The last release to remove non-@Beta APIs was Guava 21.0.) Even @Deprecated APIs will remain (again, unless they are @Beta). We have no plans to start removing things again, but officially, we’re leaving our options open in case of surprises (like, say, a serious security problem).

I think it's fine to 1)use vanilla Guava jar directly, 2) avoid using Guava @Beta APIs in Spark, 3)keep Guava in the latest version.

Comment (sunchao) [reply dev/deps/spark-deps-hadoop-2.7-hive-2.3:62]: Thanks for the reference! this is good from Spark side but it won't help if downstream apps use beta API which were changed and thus causing conflicts. I agree with you that we should avoid using these beta APIs - perhaps worth adding a maven plugin to check this.

In the mean time I'll leave the guava dependency in the assembly module and we should revisit it separately. Ideally I feel we should remove it if only `curator-client` is blocking it but we'll need to double check on that.

File: dev/deps/spark-deps-hadoop-2.7-hive-2.3
Comment (sunchao) [inline dev/deps/spark-deps-hadoop-2.7-hive-2.3:60]: I'm not sure why these datanucleus jars are removed here - they are still used by the `hive` module and present in the Spark distribution jars.

**Comment by HyukjinKwon** [comment]
cc @srowen too FYI

File: sql/hive-shaded/pom.xml
Comment (pan3793) [inline sql/hive-shaded/pom.xml:64]: Can the property `${hive.group}` be replace with literal `org.apache.hive`? I also see there are `<groupId>org.apache.hive</groupId>` somewhere.

File: sql/hive-shaded/pom.xml
Comment (pan3793) [inline sql/hive-shaded/pom.xml:93]: This means that the shaded jar will be different when packaging w/ or w/o `-Phive-thriftserver`, which one will be published to maven central?

File: sql/hive-shaded/pom.xml
Comment (pan3793) [inline sql/hive-shaded/pom.xml:198]: `javax.jdo`, `javax.realtime`, `javax.transaction`, ``javax.xml`, `org.json` still exist in the `spark-hive-shaded` jar with the original package name.

File: sql/hive-shaded/pom.xml
Comment (pan3793) [inline sql/hive-shaded/pom.xml:137]: It includes some log conf or other conf files into shaded jar, like `bonecp-default-config.xml`, `hive-exec-log4j2.properties`, `hive-log4j2.properties`, `package.jdo`, `parquet-logging.properties`, `testpool.jocl`, `tez-container-log4j2.properties`, does spark really need them?

File: sql/hive/pom.xml
Comment (pan3793) [inline sql/hive/pom.xml:131]: Why handle `datanucleus` separately?

File: sql/hive/pom.xml
Comment (pan3793) [inline sql/hive/pom.xml:132]: Why handle `datanucleus` separately?

File: sql/hive/pom.xml
Comment (pan3793) [inline sql/hive/pom.xml:162]: ditto, why handle `jline` separately?

Comment (sunchao) [reply sql/hive/pom.xml:162]: Spark already declared `jline` dependency (with the same version as that from Hive) so I excluded it from the shaded jar and added it here as a transitive dependency of `spark-hive` module.

File: sql/hive-shaded/pom.xml
Comment (sunchao) [reply sql/hive-shaded/pom.xml:198]: Oops forgot these. I'm thinking to add these `javax.*` dependencies explicitly in `spark-hive` and exclude them from the uber jar, since they seem relatively stable. What do you think?

On the other hand, seems we should exclude `org.json` from Hive due to SPARK-18262.

File: sql/hive/pom.xml
Comment (pan3793) [reply sql/hive/pom.xml:132]: Does spark use datanucleus directly?

File: sql/hive/pom.xml
Comment (pan3793) [reply sql/hive/pom.xml:162]: Does spark use datanucleus directly? If no, I think we can put it into uber jar with relocation.

File: sql/hive/pom.xml
Comment (sunchao) [reply sql/hive/pom.xml:132]: No, but I think it's a bit risky to relocate datanucleus. If you look at [HiveUtils](https://github.com/apache/spark/blob/master/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala#L518) there are places where class name from datanucleus is directly referenced. After relocation it may not function properly.

File: sql/hive-shaded/pom.xml
Comment (sunchao) [reply sql/hive-shaded/pom.xml:64]: I'm fine with either way - don't see much difference between the two.

**Comment by pan3793** [comment]
Overall this PR LGTM, and left few comments. It freed Spark from Hive's dependency hell, which is a major improvement in Spark dependency management.

Another question, the `build/mvn` command works well with the new module `spark-hive-shaded`, but my IDEA does not seem to recognize the classes from the `spark-hive-shaded` jar.

File: sql/hive/pom.xml
Comment (pan3793) [reply sql/hive/pom.xml:132]: Got it. Thanks for the explanation.

File: sql/hive-shaded/pom.xml
Comment (pan3793) [reply sql/hive-shaded/pom.xml:93]: Got it. Thanks for the explanation.

File: assembly/pom.xml
Comment (dongjoon-hyun) [inline assembly/pom.xml:83]: Could you file a JIRA and make this an IDed TODO like `TODO(SPARK-XXX)`?

File: dev/deps/spark-deps-hadoop-2.7-hive-2.3
Comment (dongjoon-hyun) [inline dev/deps/spark-deps-hadoop-2.7-hive-2.3:30]: In general, for all new binaries, we need to check the binary license and update `LICENSE-binary`.
If we don't use this, can we cut this transitive dependencies?

File: dev/deps/spark-deps-hadoop-2.7-hive-2.3
Comment (dongjoon-hyun) [inline dev/deps/spark-deps-hadoop-2.7-hive-2.3:60]: ditt. `LICENSE-binary`.

File: dev/deps/spark-deps-hadoop-3.2-hive-2.3
Comment (dongjoon-hyun) [inline dev/deps/spark-deps-hadoop-3.2-hive-2.3:54]: Do you know why this PR upgrade GSON in `Hadoop 3.2 + Hive 2.3` profile only?

File: sql/hive-shaded/pom.xml
Comment (dongjoon-hyun) [inline sql/hive-shaded/pom.xml:85]: Do you think we can declare another property like `hive.guava.version` after `guava.version` in `pom.xml`?

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you, @sunchao . Only a few minor comments like new dependencies and their licenses.

**Review by viirya — COMMENTED** [review]
One question. By doing this, does the existing profile `hive-provided` still work?

**Comment by sunchao** [comment]
> Another question, the build/mvn command works well with the new module spark-hive-shaded, but my IDEA does not seem to recognize the classes from the spark-hive-shaded jar. Does IDEA work well for you?

Good catch @pan3793. I forgot to add the following to the hive shaded pom:
```
<createSourcesJar>true</createSourcesJar>
<shadeSourcesContent>true</shadeSourcesContent>
```
It seems IDEA doesn't have good support for resolving shaded classes, and I have to refer to the workaround in [here](https://youtrack.jetbrains.com/issue/IDEA-126596), by ignoring the `hive-shaded` project. But let me see if I can find a better way to do this.

**Comment by sunchao** [comment]
> One question. By doing this, does the existing profile hive-provided still work?

@viirya good catch, I forgot the `hive-provided` profile. I think it should still apply to the shaded jar but let me try it out. When the dependency is of provided scope, the shaded jar should not include the classes from it.

When the profile is activated, I guess we'll expect users to provide their own Hive jars and they'll be on their own for any potential Guava conflicts.

File: assembly/pom.xml
Comment (sunchao) [reply assembly/pom.xml:83]: will do

File: dev/deps/spark-deps-hadoop-3.2-hive-2.3
Comment (sunchao) [reply dev/deps/spark-deps-hadoop-3.2-hive-2.3:54]: I think this is because `hadoop-common` has gson dependency.

File: dev/deps/spark-deps-hadoop-2.7-hive-2.3
Comment (sunchao) [reply dev/deps/spark-deps-hadoop-2.7-hive-2.3:30]: sure, let me update these.

File: sql/hive-shaded/pom.xml
Comment (sunchao) [reply sql/hive-shaded/pom.xml:85]: good idea, let me add `hive.guava.version`

File: assembly/pom.xml
Comment (HyukjinKwon) [inline assembly/pom.xml:177]: @sunchao sorry if I missed sth but why should redeclare here?

**Comment by sunchao** [comment]
retest this please

File: assembly/pom.xml
Comment (sunchao) [inline assembly/pom.xml:172]: @viirya with this, `-Phive-provided` is now taking effect on `hive-shaded` jar. If this is turned on, the Spark binary distribution will not contain the jar and users should provide the individual Hive jars instead. They are also on their own for any Guava conflicts.

**Comment by JoshRosen** [comment]
I think we also need to bump Guava in `SparkBuild.scala`'s `DependencyOverrides`:

https://github.com/apache/spark/blob/1f40937789d14a36709b9a87238b9945ef38f1a9/project/SparkBuild.scala#L671-L672

File: sql/hive-shaded/pom.xml
Comment (JoshRosen) [inline sql/hive-shaded/pom.xml:38]: After https://github.com/apache/spark/pull/34085 lands I think we'll be able to omit the `<optional>`s here. I previously wasn't familiar with Maven's `<optional>` dependencies, so I spotted that `createDependencyReducedPom` issue while trying to understand why they were needed in this PR.

**Comment by JoshRosen** [comment]
Do we know how the Guava version upgrade will interact with the `IsolatedClientLoader` used for metastore clients?

If I'm understanding the `IsolatedClientLoader` code correctly, it looks like Guava classes will be shared between the root classloader and the metastore classloader (i.e. the metastore client will use the Spark classloader's version of that class):

https://github.com/apache/spark/blob/ef7441bb4245e2fb08d70a8025a2c6d8bebfe75b/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala#L195-L209

It looks like we also pin Guava 14.0.1 when using Ivy to resolve metastore artifacts: 

https://github.com/apache/spark/blob/ef7441bb4245e2fb08d70a8025a2c6d8bebfe75b/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala#L132-L135

Given this, could we run risks of breakage if old Hive metastore clients receive a newer version of Guava? 

Maybe we can remove this from `isSharedClass`; I'm not sure.

File: assembly/pom.xml
Comment (viirya) [reply assembly/pom.xml:172]: okay thanks for the update!

File: sql/hive-shaded/pom.xml
Comment (sunchao) [reply sql/hive-shaded/pom.xml:38]: Nice catch! yes I think we can remove these now (after the follow-up is also done).

**Comment by sunchao** [comment]
> Do we know how the Guava version upgrade will interact with the IsolatedClientLoader used for metastore clients?

This also crossed my mind too. The related unit tests (e.g., `VersionsSuite`) are all happy at the moment but there could still be unknowns I'm afraid.

> Maybe we can remove this from isSharedClass; I'm not sure.

Yes, I'm also thinking to make guava non-shared classes here so the class loader will instead pick the ones from `hive-exec` jar (which is a fat with all the guava classes). Ideally `hive-exec` should shade Guava by itself but this is only done in the latest 2.3.x and master.

**Comment by JoshRosen** [comment]
A cross-reference for other reviewers: 

Given that `hive-exec` shades Guava in Hive 2.3.8+ (https://github.com/apache/hive/pull/1356), I was initially confused about why we needed to do our own shading in this PR: I originally thought that it was done to shade a broader set of dependencies beyond just Guava, further isolating us from future dependency conflicts. As @viirya points out at https://github.com/apache/spark/pull/29326#issuecomment-875060042, though, Spark uses the `hive-exec-core` JAR, not `hive-exec`, so Hive's Guava shading doesn't apply (hence the need to shade here).

**Comment by sunchao** [comment]
Hmm interesting. After I changed the isolated class loader to pick guava classes from Hive jars (which is of 14.0.1), tests started to fail because it now seems to uses Spark's built-in Guava which is 30.1.1-jre. This doesn't seem to make sense.

```
[error] sbt.ForkMain$ForkError: java.lang.IllegalAccessError: tried to access method com.google.common.collect.Iterators.emptyIterator()Lcom/google/common/collect/UnmodifiableIterator; from class org.apache.hadoop.hive.ql.exec.FetchOperator
[error] 	at org.apache.hadoop.hive.ql.exec.FetchOperator.<init>(FetchOperator.java:108)
[error] 	at org.apache.hadoop.hive.ql.exec.FetchTask.initialize(FetchTask.java:87)
[error] 	at org.apache.hadoop.hive.ql.Driver.compile(Driver.java:541)
[error] 	at org.apache.hadoop.hive.ql.Driver.compileInternal(Driver.java:1317)
[error] 	at org.apache.hadoop.hive.ql.Driver.runInternal(Driver.java:1457)
[error] 	at org.apache.hadoop.hive.ql.Driver.run(Driver.java:1237)
[error] 	at org.apache.hadoop.hive.ql.Driver.run(Driver.java:1227)
[error] 	at org.apache.spark.sql.hive.client.HiveClientImpl.$anonfun$runHive$1(HiveClientImpl.scala:831)
```

`Iterators.emptyIterator` here is no longer public in the newer versions of guava.

**Comment by JoshRosen** [comment]
I think the test failures seen above are due to the tests not picking up the shaded classes: I wouldn't expect to see a reference to non-relocated Guava classnames if shaded Hive was being used by the tests.

The SBT build doesn't perform any shading (because the `sbt-pom-reader` plugin only captures the project dependency structure, not build plugins). To date this hasn't been a problem since Spark's existing shading was only for the benefit of downstream consumers of Spark itself and wasn't necessary for Spark's internal functioning (and official releases are published with Maven, not SBT). With this new `hive-shaded`, though, we'll need to find a way to wire shaded classes into the SBT build.

I think we'll also run into similar problems in the Maven build. According to [Maven's build lifecycle docs](https://maven.apache.org/guides/introduction/introduction-to-the-lifecycle.html):

> `test` - test the compiled source code using a suitable unit testing framework. These tests should not require the code be packaged or deployed

The shading runs during the `package` phase, which is after the test phase, so the tests won't be able to use the shaded artifacts.

The `hive-shaded` module doesn't depend on any source code from Spark. Given this, a clean path forward might be to publish this Maven artifact separately from Spark's main build, then depend on the published artifact. This would be similar to the approach taken in https://github.com/apache/hbase-thirdparty and https://github.com/apache/hadoop-thirdparty . This would require a bunch of new infra setup, though, so if we're considering pursuing this then we should open a new thread on the dev@ mailing list to discuss.

For now, we can test out that approach via manual local publishing: that'd unblock progress on testing the `IsolatedClientLoader` and other things.

Why did the tests pass before? It looks like your most recent commit in c2ec1f8122ea9b871a1679571f0b20009be39cb7 included the `DependencyOverrides` version bump in SparkBuild.scala as well as the change to the `IsolatedClientLoader`. Prior to the `DependencyOverrides` change, the SBT build was using both unshaded Hive and old Guava, so I think the previous successful tests were passing accidentally and that the new failures aren't caused by the `IsolatedClientLoader` change.

**Comment by JoshRosen** [comment]
Also, I spent a bit of time re-reading the [`IsolatedClientLoader` code](https://github.com/apache/spark/blob/c2ec1f8122ea9b871a1679571f0b20009be39cb7/sql/hive/src/main/scala/org/apache/spark/sql/hive/client/IsolatedClientLoader.scala#L249-L277) and think I now have a better understanding of how it works and have a better explanation for why I think we should exclude Guava from `sharedClasses`:

---

First, let's consider the case where the metastore and execution Hive versions are the same:

In the `else if (!isSharedClass(name))` branch we'll try to load from the isolated classloader itself via a `super.loadClass` call. This will first attempt to load classes from `rootClassLoader` and then will fall back on loading classes from the jars specified in `allJars`. When the built-in execution Hive is used, it [looks like](https://github.com/apache/spark/blob/c2ec1f8122ea9b871a1679571f0b20009be39cb7/sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveUtils.scala#L385-L403) `allJars` gets populated from the context/Spark classloader's JARs (which is effectively the classpath used by Spark user code). In this case we'd expect Guava 30.1.1-jre to be used.

In the `else` branch (taken when `isSharedClass(name) == true`), we first try to load from `baseClassLoader`, which is the context/Spark classloader. Failing that, we then try try the steps described in the previous branch.

This makes sense to me: if the metastore and execution Hive versions are the same then there's no separate classpath of downloaded or user-provided metastore JARs to use, so we'll wind up loading classes using Spark's own classpath. In this case I think the only difference in these two branches is whether we'll use already-loaded classes from Spark's classloader or whether we'll load them again in the IsolatedClientLoader, but in both cases the classpath should be the same. If Spark's Guava version is the same as the version used by Hive then it's preferable to load from the shared classloader so that we avoid unnecessary duplicate classloading.

---

What happens when the metastore Hive version is **not** the same as the execution Hive version?

In that case, non-shared classes will be loaded from `allJars`, which should be a different classpath from Spark itself. As a result, I think we should exclude Guava from the shared classes. 

If Spark resolves the metastore JARs from Maven, this `allJars` path should include a downloaded copy of Guava 14.0.1. 

A common pattern that I've seen is that users will use Spark's existing Maven-based metastore resolution once to download the JARs, then will copy the result to a durable location to avoid re-downloading in the future. In that case they'll already have the required old version of Guava and things should continue to work while upgrading.

There's a tricky corner-case if a user has manually built a metastore classpath which includes only the dependencies not already provided by Spark: in that case, an upgrade might cause the user to run into dependency errors if Spark upgrades its Guava and the user's metastore JARs path doesn't include the old version. This is a purely hypothetical scenario and I imagine it would be rare, but it's potentially worth flagging in the docs' upgrade guide.

**Comment by sunchao** [comment]
Thanks @JoshRosen ! These are some great analysis!

> I think we'll also run into similar problems in the Maven build. According to Maven's build lifecycle docs:

I've completely missed this 🤦 . Yes adding the `hive-shaded` module in Spark will not be a good idea given the above reasons on SBT and Maven test lifecycle, and now I understand why other projects put the shaded library in a different repo :)

Let me spend more time to revisit the following two paths:
1. shade all the dependencies in Hive (e.g., via `hive-exec` fat jar) and make a new release, so Spark can start using that.
2. create a ASF repo such as `spark-thirdparty` following the examples from HBase & Hadoop. This needs community discussion as you mentioned, and I'm not sure how much more burden it will add to Spark's maintenance procedure.

> There's a tricky corner-case if a user has manually built a metastore classpath which includes only the dependencies not already provided by Spark

Thanks for the detailed explanation on how the `IsolatedClientLoader` works, and I agree this is a minor issue we should be aware of. We can either put something on the release notes, or perhaps exclude unshaded Guava jar completely from the Spark distribution (for `hadoop-3.2`). Currently this appears to be blocked by the `curator-client` dependency as discussed earlier in the PR, and perhaps there is still a way to ship only shaded Guava (from `network-common`) with those few classes required by `curator-client` excluded from relocation.

> One more consideration: what about Hadoop 2.7 builds?

Another good question :) You are right that Hadoop 2.7 still uses unshaded Guava, while Hadoop 3.3.1 has switched to use shaded Guava via HADOOP-17288. In addition Spark is using shaded Hadoop client from HADOOP-11804 which further relocates other Hadoop dependencies so they won't pollute Spark's classpath. 

I think one approach is to keep Guava 14.0.1 for `hadoop-2.7` profile so everything still stay the same there. This at least will unblock us from upgrading Guava for the default `hadoop-3.2` profile, and make sure all the published Spark artifacts will get the newer version of Guava. Also the aforementioned idea of excluding unshaded Guava from Spark distribution will only apply for the latter.

A crazier idea is to shade Hadoop 2.7 also if we are going with the `spark-thirdparty` approach but I'm not sure whether it's worth it given we are going to deprecate `hadoop-2.7` eventually.

**Pull request closed without merging**
